Player-controlled Regal Rats now lick or scratch based on their combat mode

atosti opened the pull request:
## About The Pull Request
This updates Regal Rats to now check for a user's combat mode when determining whether to lick or scratch a living target (Lick = combat off, and Scratch = combat on). It also Closes #68621.

Additionally, it allows for player movement during the licking interaction to properly cancel that action. Previously, it would attempt a lick, the player would cancel it, and then the player would attack their target instead (this was due to an oversight in the code).

Edit: It now also prevents NPC Regal rats from licking entirely, and fixes their logic so that they move on to a new target after completing a kill.

## Why It's Good For The Game
It allows players to control what their Regal rat does according to their combat mode, which matches the behavior for other player-controlled creatures (such as humans). Plus, this allows Regal rats control over when to harm living creatures, making them much less vulnerable than they were before.

Edit: As for NPC Regal rats, it allows them to be a bit more dangerous, since they'll now only attack creatures that are alive. Moreover, removing the ability to lick from NPCs helps avoid the strange scenario where players could belong to a faction which is led by an NPC. (How can one serve a King that lacks a sentient mind? Spacemen aren't ready for that type of philosophical dilemma.)

## Changelog
:cl:
fix: Player-controlled Regal Rats now lick or claw based on their combat mode.
fix: NPC Regal rats now properly claw at enemies, moving on to new targets after completing kills.
/:cl:

atosti committed:
Regal rats can now lick or scratch living targets based on their combat mode

atosti commented:
There is a side-note that I want to add to this PR:

These changes also adjust the logic for licking/scratching in NPC Regal rats (because it's shared code). NPC Regal rats used to be hostile by licking a target indefinitely, but could scratch the target if it attempted to move away from them (due to a coding oversight). I have fixed this oversight, but that means these rats will now ONLY lick you to death.

Although this is still somewhat harmful, it feels a bit silly that they never scratch living creatures. It feels like I should add some sort of logic for mixing up their attacks (e.g. have 25% of their attacks are licks and 75% are slashes), but that begins to become a balancing change rather than simply a bug fix.

I'm more than happy to attempt an implementation of something like that in this PR, but would like guidance on a suggested implementation if so. Otherwise, I think a new issue should be opened addressing that NPC Regal rats need adjustment to their internal attack logic.

Helg2 commented:
> These changes also adjust the logic for licking/scratching in NPC Regal rats (because it's shared code). NPC Regal rats used to be hostile by licking a target indefinitely, but could scratch the target if it attempted to move away from them (due to a coding oversight). I have fixed this oversight, but that means these rats will now ONLY lick you to death.

I hope you know that licking provides rat_spit which if overdosed will add you to rat_faction. And if you're making it only lick you and not scratch, its basically almost free chance to kill somebody if they attack rat king, and it's quite strange that you can just attack and run away tactic the rat king and it won't do anything else to you.

atosti commented:
@Helg2 To be clear, what you're describing is basically how the rat king currently works, even without these changes. The only time it attacks you is when you move away from being adjacent to it. If you stand next to it, you can just punch it freely without being hit (but licked relentlessly).

In relation to the rat faction, I think that brings up a good point about what the intended AI for the Regal Rat should be (or was). It's unclear to me whether NPC Regal rats were ever intended to recruit into their faction via licking. If not, then maybe just having them claw at enemies when provoked makes the most sense.

atosti committed:
NPC regal rats no longer lick targets and will now move on after completing a kill

atosti committed:
Merge branch 'tgstation:master' into regal-rat

atosti commented:
I've adjusted NPC Regal rats to now only claw their targets (no licking) and also fixed their target selection to move on after completing a kill. The PR body has been updated to reflect these changes.

atosti committed:
Merge branch 'master' into regal-rat

atosti committed:
Merge branch 'master' into regal-rat

atosti committed:
Merge branch 'master' into regal-rat

san7890 commented on code/modules/mob/living/simple_animal/hostile/regalrat.dm:
```diff
@@ -117,27 +118,19 @@
 		return
-
-	if (target.reagents && target.is_injectable(src, allowmobs = TRUE) && !istype(target, /obj/item/food/cheese))
+	if (target.reagents && target.is_injectable(src, allowmobs = TRUE) && !istype(target, /obj/item/food/cheese) && !src.combat_mode && src.mind)
```
this short-circuits, so re-arrange the order so that we can get the simplest (or those most likely to be falsey) checks out of the way the quickest. something like:

```suggestion
	if (src.mind && !src.combat_mode && target.reagents && target.is_injectable(src, allowmobs = TRUE) && !istype(target, /obj/item/food/cheese))
```

atosti replied:
That's a good call, tbh, I'll keep it in mind for future commits as well.

atosti committed:
Rearranges order to check most likely outcomes first

Co-authored-by: san7890 <the@san7890.com>

atosti committed:
Merge branch 'master' into regal-rat
